### PR TITLE
[CARBONDATA-3056] Added concurrent reading through SDK

### PR DESCRIPTION
**Pull request by NamanRastogi**
Added an API  _CarbonReader.split_ to enable concurrent reading of carbondata files through SDK.
```java
List<CarbonReader> multipleReaders = CarbonReader.split(maxSplits)
```

For detailed information on how to use this API for concurrent reading, please refer **ConcurrentSdkReaderTest.java**

## Performance Metrics:

| | configured table block: 1 MB | configured table block size: 10 MB | configured table block: 100 MB |
| --- | --- | --- | --- |
| **# rows: 1e6**<br>**Store: 7.6 MB** | # files generated: 11<br><br>Sequential Read: 274 ms<br>Parallel Read: 123 ms | # files generated: 1<br><br>Sequential Read: 247 ms<br>Parallel Read: 248 ms | # files generated: 1<br><br>Sequential Read: 252 ms<br>Parallel Read: 254 ms |
| **# rows: 1e7**<br>**Store: 78 MB** | # files generated: 104<br><br>Sequential Read: 2685 ms<br>Parallel Read: 1230 ms | # files generated: 9<br><br>Sequential Read: 2499 ms<br>Parallel Read: 1357 ms | # files generated: 1<br><br>Sequential Read: 2527 ms<br>Parallel Read: 2597 ms |
| **# rows: 1e8**<br>**Store: 865 MB** | | # files generated: 95<br><br>Sequential Read: 27069 ms<br>Parallel Read: 16082 ms | # files generated: 15<br><br>Sequential Read: 25841 ms<br>Parallel Read: 13256 ms |

 - [ ] Any interfaces changed?
 - [x] Any backward compatibility impacted: No
 - [ ] Document update required?
 - [x] Testing done
        - New unit test case have been added.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Can one of the admins verify this patch?

**Comment by xuchuanyin** [comment]
@NamanRastogi I think we can further optimize this function.

1. We can enable the parallel reading and set the parallelism while creating a CarbonReader;
2. Inside CarbonReader, we handle the concurrent processing;
3. The interfaces for CarbonReader should be kept the same as before, there is no need to add more interfaces. By calling hasNext or next, user can get the next record and will not care about which RecordReader does this record belong to.

The user interface looks like below:
```
CarbonReader reader = CarbonReader.builder(dataDir).parallelism(3).build();
while (reader.hasNext()) {
  reader.next();
}
reader.close();
```
To keep it simple, by default the parallelism can be 1 which means we will process each RecordReader one by one. Setting this parallelism to a higher value means that we will go process the RecordReaders in a thread pool with size 3.

**Comment by NamanRastogi** [comment]
@xuchuanyin 

Since `CarbonReader` is iterator based, we can only read a line when user wants. So, even if we keep the parallelism internally before the build method, it will still read the file row by row (even though they are in different threads) and reading multiple rows will still be accessed sequentially. Please notice the test file **ConcurrentSdkReaderTest.java**, the reading is happening in the thread itself, and multiple threads are reading the files (line-by-line) concurrently.

The actual reading is happening inside the `CarbonReader.readNext()` method, as long as that is sequential, the actual reading performance is not going to get better, so we have to have `CarbonReader.readNext()` inside different threads,.

**Comment by xuchuanyin** [comment]
emm, but in your implementation, most of the work has to be done by the user (multi-thread handling). CarbonData itself only split the input data and return multiple readers. If this is the solution, why not just tell the user to generate multiple CarbonReaders by passing only part of the input dir each time they create the reader?

Addition to my proposal, I think we can add a buffer for the records. When `CarbonReader.next` is called, we can retrieve the record from the buffer and fill the buffer asynchronously. When`CarbonReader.hasNext` is called, we can first detect this from the buffer, if it is empty, we will then detect this from the recordReader and fill the buffer asynchronously.

**Comment by NamanRastogi** [comment]
@xuchuanyin 
Using this API the overhead of reading concurrently is on the consumer of CarbonReader. So yes, what you said what you said is right. But this API was made to consider this case only when a user wants to have more manual control over the concurrent reading. Consider the scenario when the user wants to read different files in different machines. If we take care of the concurrent reading internally (using a buffer as you suggested), the user cannot do that.

Another thing that you mentioned was user providing the location of _carbondata_ files in the SDK reader itself, and that does not work in the current implementation of SDK.

And for faster reading, there is one more pull request: [2816](https://github.com/apache/carbondata/pull/2816), it supports batch reading for a better reading performance. It is still not concurrent reading, but better than iterator based row-by-row reading.

File: store/sdk/src/test/java/org/apache/carbondata/sdk/file/ConcurrentSdkReaderTest.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+import org.junit.*;
+
+/**
+ * multi-thread Test suite for {@link CarbonReader}
+ */
+public class ConcurrentSdkReaderTest extends TestCase {
+
+  private static final String dataDir = "./testReadFiles";
+
+  public void cleanTestData() {
```
Comment (xubo245) [inline store/sdk/src/test/java/org/apache/carbondata/sdk/file/ConcurrentSdkReaderTest.java]: you can add @Before or @After got it. especially there are many test case

File: store/sdk/src/test/java/org/apache/carbondata/sdk/file/ConcurrentSdkReaderTest.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+      count = 0;
+      long start = System.currentTimeMillis();
+      for (CarbonReader reader_i : multipleReaders) {
+        results.add(executorService.submit(new ReadLogic(reader_i)));
+      }
+      for (Future result_i : results) {
+        count += (long) result_i.get();
+      }
+      long end = System.currentTimeMillis();
+      System.out.println("[Parallel read] Time:" + (end - start));
```
Comment (xubo245) [inline store/sdk/src/test/java/org/apache/carbondata/sdk/file/ConcurrentSdkReaderTest.java]: Please add unit for it, such as ms

**Comment by xubo245** [comment]
@NamanRastogi Hi, customer required carbon provide the same order between using one thread to read and use multiple threads to read data.

**Comment by NamanRastogi** [comment]
Yes, data coming from one file will always be in order. Please check the `split` method, it splits the list of CarbonRecordReader into multiple CarbonReader s.

Suppose there are 10 carbondata files, and the user wants to get 3 splits, so he will get a list like this:

**Comment by NamanRastogi** [comment]
Please check the split method, it splits the list of `CarbonRecordReader` into multiple `CarbonReader`s. It does not jumble the order of `CarbonRecordReader`, it still keeps them sequential.

Suppose there are 10 *carbondata* files and thus 10 `CarbonRecordReader` in `CarbonReader.readers` object and the user wants to get 3 splits, so he will get a list like this:
```java
CarbonReader reader = CarbonReader.builder(dataDir).build();
List<CarbonReader> multipleReaders = reader.split(3);
```
And the indices of `CarbonRecordReader`s in `multipleReaders` will be like:
`multipleReaders.get(0).readers` points to {0,1,2,3} indices of *carbondata* files
`multipleReaders.get(1).readers` points to {4,5,6} indices of *carbondata* files
`multipleReaders.get(2).readers` points to {7,8,9} indices of *carbondata* files

Now, if you read the rows like following code, the rows will still be in order.
```java
for (CarbonReader reader_i : multipleReaders) {
    reader_i.readNextRow();
}
```

Earlier, you were getting data from 5th `CarbonRecordReader` only after you have exhausted the 4th. But now, you are getting it earlier, maybe even before 0th. So the user has to make sure he consumes it after he has used up the 4th file if order is important for him/her, otherwise he/she can use it earlier also if order is not important. So, for example to count the total no. of rows, user does not need the original order.

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java [hunk 3]
```diff
@@ -114,6 +117,43 @@ public static CarbonReaderBuilder builder(String tablePath) {
     return builder(tablePath, tableName);
   }
 
+  /**
+   * Return a new list of {@link CarbonReader} objects
+   *
```
Comment (ajantha-bhat) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java]: Add a clear description, mention what happens if splits greater than the number of files and what happens if splits are lesser than the number of files

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java [hunk 4]
```diff
@@ -114,6 +117,43 @@ public static CarbonReaderBuilder builder(String tablePath) {
...
+
+    List<CarbonReader> carbonReaders = new ArrayList<>();
+
+    // If maxSplits < readers.size
+    // Split the reader into maxSplits splits with each
+    // element contains >= 1 CarbonRecordReader objects
+    if (maxSplits < this.readers.size()) {
+      for (int i = 0; i < maxSplits; ++i) {
+        carbonReaders.add(new CarbonReader<>(this.readers
+            .subList((int) Math.ceil((float) (i * this.readers.size()) / maxSplits),
```
Comment (ajantha-bhat) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java]: this is constant, do this outside loop and use it each time

File: store/sdk/src/test/java/org/apache/carbondata/sdk/file/ConcurrentSdkReaderTest.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+      }
+      writer.close();
+    } catch (Exception e) {
+      e.printStackTrace();
+      Assert.fail(e.getMessage());
+    }
+  }
+
+  @Test public void testReadParallely() throws IOException, InterruptedException {
+    long numRows = 10000000;
```
Comment (ajantha-bhat) [inline store/sdk/src/test/java/org/apache/carbondata/sdk/file/ConcurrentSdkReaderTest.java]: We must not add huge record test case in UT. PR builder time for all the PR will affect by this. Locally test with huge data but update UT with fewer rows. say 10 rows.

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java (hunk 3 above)
Comment (NamanRastogi) [reply store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java]: Done!

**Comment by shardul-cr7** [comment]
retest this please

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java [hunk 6]
```diff
@@ -114,6 +115,57 @@ public static CarbonReaderBuilder builder(String tablePath) {
...
+   * through >= 1 file.
+   *
+   * If the no. of files is less than maxSplits, then return list of
+   * CarbonReader with size as the no. of files, with each CarbonReader
+   * iterating through exactly one file
+   *
+   * @param maxSplits: Int
+   * @return list of {@link CarbonReader} objects
+   */
+  public List<CarbonReader> split(int maxSplits) throws IOException {
```
Comment (ravipesala) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java]: I feel this method should be moved to builder. Add another method in builder `build(int splits)` and return List of readers.

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java [hunk 7]
```diff
@@ -114,6 +115,57 @@ public static CarbonReaderBuilder builder(String tablePath) {
...
+    validateReader();
+    if (maxSplits < 1) {
+      throw new RuntimeException(
+          this.getClass().getSimpleName() + ".split: maxSplits must be positive");
+    }
+
+    List<CarbonReader> carbonReaders = new ArrayList<>();
+
+    if (maxSplits < this.readers.size()) {
```
Comment (ravipesala) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java]: Add UT only to this method to make sure splits happen correctly with multiple splits combinations and readers size,

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java (hunk 6 above)
Comment (ajantha-bhat) [reply store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java]: @ravipesala : Adding to builder will break builder pattern, recently we removed arguments from build() and make it as separate API for SDK writer. Reader also followed same.

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java (hunk 7 above)
Comment (kunal642) [reply store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java]: @ravipesala Let us add test cases in a separate PR. would it be okay?

**Comment by ravipesala** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1244/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9509/

**Comment by kunal642** [comment]
LGTM

**Pull request closed without merging**